app/src/org/commcare/views/widgets/CalendarFragment.java

```diff
@@ -0,0 +1,342 @@
+            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
+                calendar.set(Calendar.MONTH, position);
+                refresh();
+            }
+
+            @Override
+            public void onNothingSelected(AdapterView<?> parent) {}
+        });
+    }
+
+    private void initOnClick(){
+
+        myGrid.setOnItemClickListener(new AdapterView.OnItemClickListener() {
+            @Override
+            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
+                Date date = (Date) parent.getItemAtPosition(position);
+                calendar.setTime(date);
+                refresh();
+            }
+        });
+
+        cancel.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                dismiss();
+                if(myListener != null){
+                    myListener.onCalendarCancel();
+                }
+            }
+        });
+
+        ImageButton closer = (ImageButton) myLayout.findViewById(R.id.close_calendar);
+        closer.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                dismiss();
+                if(myListener != null){
+                    myListener.onCalendarClose();
+                }
+            }
+        });
+
+        today.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                calendar.setTimeInMillis(todaysDateInMillis);
+                refresh();
+            }
+        });
+    }
+
+    //Redraws the calendar display
+    public void refresh(){
+        ArrayList<Date> dateList = new ArrayList<>();
+        Calendar populator = (Calendar) calendar.clone();
+
+        int totalDays = getNumDaysInMonth(populator);
+        populateListOfDates(dateList, populator, totalDays);
+
+        yearSpinner.setSelection(calendar.get(Calendar.YEAR)-GregorianDateWidget.MINYEAR);
+        monthSpinner.setSelection(calendar.get(Calendar.MONTH));
+        myGrid.setAdapter(new CalendarAdapter(getContext(), dateList));
+    }
+
+    //Populates an arraylist with dates
+    private void populateListOfDates(ArrayList<Date> dateList, Calendar populator, int totalDays) {
+        while(dateList.size() < totalDays){
+            dateList.add(populator.getTime());
+            populator.add(Calendar.DAY_OF_MONTH, 1);
+        }
+
+        int remainingDays = ((DAYSINWEEK + 1)-(populator.get(Calendar.DAY_OF_WEEK)))%DAYSINWEEK;
+
+        for(int i = 0; i < remainingDays; i ++){
+            dateList.add(populator.getTime());
+            populator.add(Calendar.DAY_OF_MONTH, 1);
+        }
+    }
+
+    //Calculates the total number of days from most recent Sunday to the first Saturday following the end of the current month to fill the calendar view
+    private int getNumDaysInMonth(Calendar cal){

```

Comment:
Marking methods as `static` is important because they are easier refactor, unit test, and very slightly faster to invoke